### PR TITLE
gh-125206: Correct detection of complex numbers support in libffi

### PR DESCRIPTION
**Pull request by skirpichev**
* Issue: gh-125206

### CODE REVIEW

**Comment by skirpichev** [comment]
!buildbot SPARCv9 Oracle Solaris 11.4

**Comment by skirpichev** [comment]
!buildbot PPC64LE CentOS9 LTO + PGO

**Comment by skirpichev** [comment]
!PPC64LE CentOS9 LTO + PGO PR

**Comment by skirpichev** [comment]
!builbot PPC64LE CentOS9 LTO + PGO PR

**Comment by skirpichev** [comment]
!buildbot PPC64LE CentOS9 LTO + PGO PR

**Comment by skirpichev** [comment]
!buildbot PPC64LE CentOS9 LTO \+ PGO

File: configure.ac [hunk 1]
```diff
@@ -4098,8 +4098,10 @@ AS_VAR_IF([have_libffi], [yes], [
 # See https://github.com/python/cpython/issues/125206 for more details.
 #
 AC_CACHE_CHECK([libffi has complex type support], [ac_cv_ffi_complex_double_supported],
-[ac_save_cc="$CC"
-CC="$CC -lffi"
+[WITH_SAVE_ENV([
+ CFLAGS="$LIBFFI_CFLAGS $CFLAGS"
```
Comment (erlend-aasland) [inline configure.ac]: Let's use `CPPFLAGS` instead; we've had `CFLAGS` fail mysteriously on FreeBSD earlier.

```suggestion
 CPPFLAGS="$LIBFFI_CFLAGS $CPPFLAGS"
```

Comment (vstinner) [reply configure.ac]: CPPFLAGS are flags for the C preprocessor?

Comment (erlend-aasland) [reply configure.ac]: Yes, correct; the C preprocessor will use `CPPFLAGS` (but not `CFLAGS`), and the compiler will use both `CFLAGS` and `CPPFLAGS`. Sometimes Autoconf will run parts of a check only with the pre-processor. We've had issues with this before. I think that for the following `AC_RUN_IFELSE` macro, `CFLAGS` should be fine, but I prefer to play safe and use `CPPFLAGS`.

**Comment by skirpichev** [comment]
PPC64LE build works as expected (no support)

File: configure.ac (hunk 1 above)
Comment (vstinner) [reply configure.ac]: The error https://github.com/python/cpython/issues/125206#issuecomment-2443158799 may be caused by missing *linker* flags, LDFLAGS?

File: configure.ac [hunk 2]
```diff
@@ -4098,8 +4098,10 @@ AS_VAR_IF([have_libffi], [yes], [
...
+ LDFLAGS="$LIBFFI_LIBS $LDFLAGS"
+ LIBS="$LIBFFI_LIBS"
```
Comment (erlend-aasland) [inline configure.ac]:
```suggestion
 LIBS="$LIBFFI_LIBS $LIBS"
```

File: configure.ac (hunk 1 above)
Comment (skirpichev) [reply configure.ac]: I'll do, JFR above usage was taken from the libffi test.

File: configure.ac (hunk 2 above)
Comment (skirpichev) [reply configure.ac]: Hmm, which library we need else to run the test code?

Comment (erlend-aasland) [reply configure.ac]: I would not expect any, but some build configurations are weird (cross-compile, etc.).

**Comment by skirpichev** [comment]
!buildbot SPARCv9 Oracle Solaris 11.4

**Comment by skirpichev** [comment]
Hmm, now on SPARCv9 - libffi support was detected.  Test failures seems unrelated and build was successful.  (C.f. https://buildbot.python.org/all/#/builders/1262/builds/13 and https://buildbot.python.org/#/builders/1262/builds/21)

**Review by vstinner — COMMENTED** [review]
@efimov-mikhail: Can you test the fix on Debian 10?

**Comment by efimov-mikhail** [comment]
> @efimov-mikhail: Can you test the fix on Debian 10?

Yes, I can. It looks like the correct fix, but it's better to test again.

**Review by efimov-mikhail — APPROVED** [review]

**Comment by efimov-mikhail** [comment]
This version of fix works fine on Debian 10.

**Review by erlend-aasland — APPROVED** [review]
LGTM (if the Sparc failure is unrelated).

**Comment by vstinner** [comment]
Merged, thanks.

> LGTM (if the Sparc failure is unrelated).

There are tons of unrelated errors/test failures on Solaris.

**Comment by skirpichev** [comment]
On SPARCv9 only configure test was checked.  It reports "yes" (supported).  But it seems that ctypes tests weren't run.  Sadly that configure script doesn't report library versions (even if this is available e.g. with ``pkg-config --modversion <lib>``.

**Comment by erlend-aasland** [comment]
> On SPARCv9 only configure test was checked. It reports "yes" (supported). But it seems that ctypes tests weren't run. Sadly that configure script doesn't report library versions (even if this is available e.g. with `pkg-config --modversion <lib>`.

We could fix that if we wanted.

**Comment by skirpichev** [comment]
> We could fix that if we wanted.

If that does make sense, I can open an issue.  But I admit, that I don't know how to solve this in general: maybe every snowflake^Wlibrary will require a special treatment.

BTW, version information is unavailable in the config.log (even for versioned checks like for libmpdec).

**Comment by erlend-aasland** [comment]
IIRC (from the top of my head), the pkg-config macro will print version information if you specify version bounds (see the sqlite3 third-party detection code in `configure.ac`). My gut feeling for implementing such a thing would be to create a wrapper for `PKG_CHECK_MODULES` and ensure that we print version info in the action-if-found block. I'm not sure if it is worth it, but it would be convenient for debugging third-party dependency detection.

**Pull request merged**